### DISCUSSION TITLE
py-[event] Syntax Changes

### DISCUSSION THREAD

**Discussion by JeffersGlass** [pyscript/pyscript#1222] [Proposals]
# Overview

This proposal defines two pieces of related syntax, which users can add as attributes on their HTML elements:

| Concept          | HTML Attribute  | Purpose                                                  |
|------------------|-----------------|----------------------------------------------------------|
| Event Handler    | `py-[event]`| Hooks browser events to Python event handler _functions_ |
| Event CodeRunner | `py-[event]-code` | Run Python code.                                  |

## What "`[event]`" Means Here

In all of the discussion that follows, the shorthand `[event]` is a stand-in for any* of the [Browser Events](https://developer.mozilla.org/en-US/docs/Web/Events). For example, an example which references `py-[event]=someFunc` could be read as `py-click=someFunc`, `py-change=someFunc` etc. 

# Examples

The following example user code demonstrates all the functionality in this proposal. They use the 'click' event, but any browser event can be used.

```html
<!--------------------- Event Handlers --------------------->
<py-script>
    def eventFunc(evt):
        print(f"This function got the event {evt}")
<py-script>
<button py-click="eventFunc">Takes an Event</button> (1)

<py-script>
    def noEventFunc():
        print("This function doesn't take an argument, and won't be passed one")
</py-script>
<button py-click="noEventFunc">Doesn't Take An Event</button> (2)

<py-script>
    class SomeClass():
        def oneFunc(self):
            print("This is the oneFunc method")
        
        def twoFunc(self, event):
            print(f"The twoFunc function got event {event}")

        @classmethod
        def threeFunc(self,event):
            print(f"Even classmethods work")

    instance = SomeClass()
</py-script>
<button py-click="instance.oneFunc">Instance Methods do work</button> (3)
<button py-click="instance.twoFunc">Instance Method gets passed event</button> (4)
<button py-click="Instance.threeFunc">Class Methods work too</button> (5)

<!--- Event Handlers Errors/Warnings --->

<py-script>
    def multipleNotAllowed(first, second):
        ... #Functions with multiple arguments are not valid Event Handlers
</py-script>
<button py-click="multipleNotAllowed">Click me to throw an error</button> (6)

<button py-click="print('hi')">Throws an error, since it doesn't evaluate to a Callable</button> (7)
<button py-click="print((">Throws an error, since evaluating this string fails</button> (8)



<!--------------------- Event CodeRuners --------------------->

<button py-click-code="print('hello world')">Click to Print</button> (9)
<button py-click-code="print(f'The event was {event}')">When evaluating, 'event' is a local variable</button> (10)

<py-script>
    def myFunc(a, b, c):
        print(f"{a} {b} {c}")
    
    ultimateAnswer = 42
</py-script>
<button py-click-code="myFunc('Jeff', 0, ultimateAnswer)">Call a Function with Arguments</button> (11)

<!--- Event CodeRuner Errors/Warnings --->

<button py-click-code="myFunc">Shows a warning, since this evaluates to a Callable</button> (12)
<button py-click-code="print((">Raises an error, since evaluating this fails</button> (13)
```

# Description of Behavior

When PyScript loads, following the execution off all \<py-script\> tags on:
  - any tags on the the page with an attribute of `py-[event]` are processed as "Python Event Handlers"
  - any tags on the page with an attribute of `py-[event]-code` are processed as "Python Event CodeRunners".

## Python Event Handler Behavior (`py-[event]`)

HTML Tags with an attribute matching `py-[event]` are processed in the following fashion. An event handler is registered for this element, hooked to the specified event. When the event is dispatched to the element, the following logic is executed:

The value of the `py-[event]` HTML attribute is retrieved; this is referred to as the `handlerNameString`. `handlerNameString` string is `eval()`'d in the global namespace, for the purposes of resolving what existing Python object (created in a PyScript tag, plugin tag, or similar) it references. If this evaluation fails/errors, an Python error is raised. If the result of this evaluation is not a Callable, an error is raised, along the lines of: _The code provided to 'py-[event]' should be the name of a function or Callable. To run an expression as code, use 'py-[event]-code'_

If the result _is_ a Callable, we inspect the object. If signature takes a single parameter, we call that Callable and pass it the `event` object passed by the event handler. If it takes zero arguments, we call it with no parameters. If it takes two or more parameters, an error is raised.

The method of inspection works "as expected" for instance/class methods that take a `self` or `cls` arguments, in that they are ignored. In the examples above, `py-click="instance.oneFunc"` will result in `oneFunc` being called with no arguments (its `self` parameter is ignored).

## Python Event CodeRunner Behavior (`py-[event]-code`)

HTML Tags with an attribute matching `py-[event]-code` are processed in the following fashion. An event handler is registered for this element, hooked to the specified event. When the event is dispatched to the element, the following logic is executed:

The value of the `py-[event]` HTML attribute is retrieved; this is referred to as the `userCode`. This code is `eval()`'d in the global namespace, with the local argument of `eval` equal to `{'event': js_event}`, where `js_event` is the event passed from the event dispatch on the JavaScript side. To be explicit - we are _not_ accessing Window.event or similarly accursed behavior; we are making use of an event passed by a JavaScript event handler.

For guiding users - if the result of this evaluation is a Callable, we show a warning, along the lines of: _The code provided to 'py-[event]-code' was the name of a Callable. Did you mean to use 'py-[event]?'_

# Future Improvements

The following aspects are improvements that would be useful, and are being deferred to future work.

## Mutation Handling

The first implementation of this proposal will only register the event handlers during the PyScript initialization phase; an improvement would be to have these event handlers registered/changed/removed whenever the `py-[event]` attribute on any HTML element is changed. This could be implemented using a [Mutation Observer](https://developer.mozilla.org/en-US/docs/Web/API/MutationObserver) or similar.

## *Improved event enumeration

Currently, we support a limited subset of events as specified in `pyscript.ts`, in a mapping between "py-" attributes and the name of the actual browser events. It would be good to expand this list. It would be better to somehow dynamically generate this list at initialization time before scanning the DOM for elements with the appropriate attributes.

**Comment by marimeireles** [comment]
>The value of the py-[event] HTML attribute is retrieved; this is referred to as the handlerNameString. handlerNameString string is eval()'d in the global namespace, for the purposes of resolving what existing Python object (created in a PyScript tag, plugin tag, or similar) it references. If this evaluation fails/errors, an Python error is raised. If the result of this evaluation is not a Callable, an error is raised, along the lines of: The code provided to 'py-[event]' should be the name of a function or Callable. To run an expression as code, use 'py-[event]-code'

I think we are not covering the case of "what if someone just executes code here" right? Do we intend to just throw a warning saying this is not the correct use case for this?

Great write-up! Thank you for doing this Jeff!

**Reply by JeffersGlass** [reply]
> Do we intend to just throw a warning saying this is not the correct use case for this?

I think so, yeah - if the user writes some code but it doesn't evaluate to a Callable, we show an error. This is case # 7 above (_I went back and numbered the examples_ 😇)

One confusing point for users, though, is that their code will appear to have "run just fine", then broken after, in the sense that the evaluation can succeed before the error is thrown. This is something I think we'll need to do a thorough job of explaining in the documentation, but I think is the right choice - if we later choose to use some other method for name resolution, like splitting on the '.' or adding the option to resolve in different modules/namespaces, the value passed to py-[event] remain the same.

**Reply by marimeireles** [reply]
IMO throwing a warning is enough... But yeah, let's also try to do a good job writing docs! :)

**Comment by FabioRosado** [comment]
Thank you for the great write up, I am finally catching up on notifications 😄 
The only downside of the `py-click-code` is that we will have to hardcode more things in our list, right? (Until we get mutation observer of similar working).

I would like to see how other event types like py-input or py-submit work with the new code (as well as with the @when decorator) since we haven't been testing events other than the `py-click` one.

Question:
Can we ship what we have and work on the `-code` later if it will delay us? What do we need to get this working?

I would like to ship the events in the next release so we can add good documentation around it because at the moment using events is very confusing 😄

**Reply by marimeireles** [reply]
>I would like to see how other event types like py-input or py-submit work with the new code (as well as with the https://github.com/when decorator) since we haven't been testing events other than the py-click one.

I don't see how this would change the current behavior of opening a box? Is that what you mean? For that to become smooth to the user exp. we need web workers.

And this `-code` discussion was a veeeeery long one 😅
I think most of the team converged towards using this. We definitely need something to differentiate the "normal function calling attached to event" case from the "we're writing code inside an HTML tag this is madness" one. 
The implementation is very simple... They're both using `eval` in the end and are capable of the same things, but ideally the user doesn't know that haha we're going to enforce as a bad practice to write code when not using the `-code` tag.

**Reply by marimeireles** [reply]
All that said, feel free to propose something different... I'm personally not loving this solution, seems like the best between all the options we thought on though.

**Reply by FabioRosado** [reply]
> I don't see how this would change the current behavior of opening a box? Is that what you mean? For that to become smooth to the user exp. we need web workers.

Not sure what you mean opening a box, the `py-input` is an event handler for the `input` HTML elements 😁

**Reply by JeffersGlass** [reply]
> The only downside of the py-click-code is that we will have to hardcode more things in our list, right? (Until we get mutation observer of similar working).

With a small change I think we can re-use the list of existing events for both `py-event` and `py-event-code`, _and_ make that list easier to extend. First, we make [`pyAttributeToEvent`](https://github.com/pyscript/pyscript/blob/43d598d9519b7fd5e9b19803c02f6898bddfaf2e/pyscriptjs/src/components/pyscript.ts#L66-L158) just a list of event names, instead of a map - let's call that just `eventNames`:

```ts
eventNames = ['click', 'keydown', 'afterprint', 'beforeprint', ...] 
```

Then the step where we iterate over the events and assign event listeners becomes something like:

```ts
for (const eventName of eventNames) {
    createElementsWithEventListeners(interpreter, eventName);
}

function createElementsWithEventListeners(interpreter: Interpreter, eventName: string) {
    for (const el of document.querySelectorAll(`[py-${eventName}]`)) {
        // Add event handlers here
    }

    for (const el of document.querySelectorAll(`[py-${eventName}-code]`)) {
        // Add codeRunner executors here
    }
...
}
```

**Reply by JeffersGlass** [reply]
> The implementation is very simple... They're both using eval in the end and are capable of the same things, but ideally the user doesn't know that haha we're going to enforce as a bad practice to write code when not using the -code tag.

Agreed with Mariana here, the implementation is pretty simple. There's a little bit of extra work for `py-event` to inspect the Callable the user provides and provide the event object if necessary, but that's in the demo we whipped up last Thursday - I can share that code this evening.
